pkg/agent/run.go

```diff
@@ -101,7 +101,15 @@ func Run(cmd *cobra.Command, args []string) {
 	}
 
 	dataGatherers := map[string]datagatherer.DataGatherer{}
-	var wg sync.WaitGroup
+	group, gctx := errgroup.WithContext(ctx)
+
+	defer func() {
+		// TODO: replace Fatalf log calls with Errorf and return the error
+		cancel()
+		if err := group.Wait(); err != nil {
+			logs.Log.Fatalf("failed to wait for controller-runtime component to stop: %v", err)
+		}
+	}()
```

Comment:
What is this doing? It seems like we now cancel immediately, and then block on `group.Wait`?

Reply:
Indeed, we cancel the context to start the shutdown and wait for the shutdown to complete using group.Wait

Reply:
An error group will always cancel its context when one of the goroutines returns an error.

For successful execution, we only reach this defer call when the following for loop ends because OneShot is true: https://github.com/jetstack/jetstack-secure/blob/7f756c3448bc6d541935d89fc3ccd1e7150d18f2/pkg/agent/run.go#L167-L181
If OneShot is false, the loop will keep looping.

Reply:
In my opinion, this information should be written down in a comment in code, since this isn't obvious.

Reply:
Sorry about that, I was reading `go func` instead of `defer func`... It now makes sense.